### ISSUE TITLE
Add posibility to set different capabilities for each browser session in Configuration class

### ISSUE THREAD

**Issue by ivandros** [selenide/selenide#1372]
## The problem
For running tests in Selenoid containers we set browser capabilities via Configuration.browserCapabilities = capabilities;
Like in this sample: https://github.com/selenide-examples/selenoid-sample/blob/main/src/test/java/org/selenide/examples/selenoid/FileDownloadTest.java

The problem is that in this case all simultaneous browser session in Selenoid have one the same set of capabilities.
Сouse: Configuration.browserCapabilities is static!

## Details
1. There is necessity to set in some browsers different settings. E.g. some tests need to run in browser with enabled mobile mode.
But now all browsers work only in desktope mode.

2. Selenoid has useful feature which allows to set throw capabilities unique name for each browser session and name for recorded video-file by each running test. (https://aerokube.com/selenoid/latest/#_special_capabilities). And we can do it unique and set via the code.
But now all browser session and videofile in Selenoid has the same name that complicates its identification.

3. Similar problems form other users:
 https://t.me/selenide_ru/702
 https://t.me/selenide_ru/2450
 https://t.me/selenide_ru/4601
 https://t.me/selenide_ru/5663
 https://t.me/selenide_ru/6503
 https://t.me/selenide_ru/9409
 https://t.me/selenide_ru/9509
 https://t.me/selenide_ru/16572
 https://t.me/selenide_ru/18035
 

## Tell us about your environment
* **Selenide Version**: 5.16.0
* **Chrome\Firefox\IE Version**: selenoid/vnc:chrome_87.0

## Code To Reproduce Issue

```java
 public class TestBase(){
   boolean isMobileDevice = false;   

   @BeforeMethod
   initBrowserCapabilities(this.getClass().getSimpleName(), isMobileDevice);
 
   @AfterMethod
    closeWebDriver();
}

    public class BrowserCapabilities {
    public void initBrowserCapabilities(String currentClassName, boolean isMobileDevice) {
       private String videoIdentification;
       videoIdentification = currentClassName + "_" + new SimpleDateFormat("yyyy-MM-dd_HH:mm:ss").format(new Date()).mp4";

            ChromeOptions chromeOptions = new ChromeOptions();
            DesiredCapabilities capabilities = new DesiredCapabilities();
            capabilities.setBrowserName("chrome");
            capabilities.setCapability("enableLog", true);

                capabilities.setCapability("enableVNC", true);
                capabilities.setCapability("version", "vnc_enabled_image");
                capabilities.setCapability("enableVideo", true);
                capabilities.setCapability("screenResolution", "1366x768");
                capabilities.setCapability("name", videoIdentification);
                capabilities.setCapability("videoName", videoIdentification);

            if (isMobileDevice)
                chromeOptions.addArguments("--user-agent=Mozilla/5.0 (Linux; Android 6.0.1; SM-G920V Build/MMB29K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.98 Mobile Safari/537.36");
            capabilities.setCapability(ChromeOptions.CAPABILITY, chromeOptions);

            Configuration.browserCapabilities = capabilities;
            Configuration.remote = "http://" + host + ":4444/wd/hub";
            open("");
}

public class ClassWithTests1 extends TestBase {
    public ClassWithTests1() {
        isMobileDevice = true;
    }
    @Test
    public void Test1() {
        log.info("Test is started");
                Assert.assertTrue(true);
    }
}

public class ClassWithTests2 extends TestBase {
    public ClassWithTests2() {
    }
    @Test
    public void Test2() {
        log.info("Test is started");
                Assert.assertFalse(false);
    }
}
```

**Comment by arturmkr** [comment]
Also, if would be useful to set configuration (like baseUrl) for each browser session.
Now Configuration.baseUrl is a static variable.

**Comment by asolntsev** [comment]
@arturmkr Then you probably don't need `Configuration.baseUrl` at all. You can just call `open` with absolute URL. Then `Configuration.baseUrl` will be ignored.

**Comment by asolntsev** [comment]
@ivandros I realized that this feature may be implemented much simpler than I thought before.

All we need is to add parameter `browserCapabilities` to method `open(url)`. 
Then you will just need to call

```java
String videoIdentification  = currentClassName + "_" + new SimpleDateFormat("yyyy-MM-dd_HH:mm:ss").format(new Date()).mp4";

ChromeOptions chromeOptions = new ChromeOptions();
chromeOptions.setCapability("name", videoIdentification);
chromeOptions.setCapability("videoName", videoIdentification);
open("", chromeOptions);
```

Right?

The only question: what should do method `open("", chromeOptions)` if there is already an opened browser from a previous test? Throw an exception? Or just log a warning?

**Comment by BorisOsipov** [comment]
@asolntsev 

>All we need is to add parameter browserCapabilities to method open(url).

even better add parameter `config` to method `open(url)`

>The only question: what should do method open("", chromeOptions) if there is already an opened browser from a previous test?

throw an exception

**Comment by bereg2k** [comment]
@asolntsev if it's technically possible, passing `Configuration` instance to `open()` is great.

Regarding the already opened browser session though... what can go wrong if we just log a warning, close the previous browser session, and open the new session with new configuration/capabilities?

**Comment by akrigator** [comment]
@ivandros as an option you may use "synchronized" for initBrowserCapabilities, but you still should call open() selenide to initialize browser session with initBrowserCapabilities configuration. I've use this method for naming moon sessions by test via custom listener:
```
public class ProdreviewWeb extends BrowserPerTest {
    @Override
    public void onTestStart(ITestResult result) {
        super.onTestStart(result);
        initBrowser(result);
    }

    private synchronized void initBrowser(ITestResult result) {
        Configuration.browser = WebConfiguration.getInstance()
                                                .getBrowser();
        Configuration.driverManagerEnabled = false;
        Configuration.browserSize = "1920x1080";
        Configuration.headless = WebConfiguration.getInstance()
                                                 .isHeadless();
        Configuration.baseUrl = WebConfiguration.getInstance()
                                                .getCommonConfiguration()
                                                .getProdRevUrl();
        Configuration.remote = WebConfiguration.getInstance()
                                               .getRemoteDriverUrl();
        Configuration.browserCapabilities.setCapability("moon:options",
                                                        Map.of("sessionTimeout", WebConfiguration.getInstance()
                                                                                                 .getRemoteDriverTimeout(),
                                                               "name", result.getMethod().getQualifiedName(),
                                                               "enableVnc", true));
        Selenide.open();
    }
}
```

And set `@Listeners(ProdreviewWeb.class)` annotation in each test classes

**Comment by asolntsev** [comment]
@bereg2k 
> @asolntsev if it's technically possible, passing `Configuration` instance to `open()` is great.

The problem is that there are different settings in `Configuration`: some are used for opening a browser, other are used later in the following Selenide work. Probably we could extract browser-related settings to a separate class, like `BrowserSettings`?.. 

> Regarding the already opened browser session though... what can go wrong if we just log a warning, close the previous browser session, and open the new session with new configuration/capabilities?

Yes, it's possible. The only drawback is that we will close the browser every time, even if the settings have not been changed.

**Comment by asolntsev** [comment]
> as an option you may use "synchronized" for initBrowserCapabilities.

@akrigator Technically it's possible, but i's rather a hack. Using `synchronized` method kills the parallelism.

**Comment by akrigator** [comment]
> > as an option you may use "synchronized" for initBrowserCapabilities.
> 
> @akrigator Technically it's possible, but i's rather a hack. Using `synchronized` method kills the parallelism.

Yeap, it's workaround for current implementation while config is static. About parallelism - yes, tests are stacked till configuration, but configuration is not a long operation in comparing with test execution, so tests run in parallel still.

In any case, it will nice to have thread local configuration, like webdriverContainer or staticSelenideDriver in WebDriverRunner

**Comment by akrigator** [comment]
> > In any case, it will nice to have thread local configuration, like webdriverContainer or staticSelenideDriver in WebDriverRunner
> 
> @akrigator Yes, I am going to implement it. But can you specify what settings from `Configuration` you want to change from test to test (except `Configuration.browserCapabilities`)?

`name` in `moon:options` from previous comment, thanks!

**Comment by asolntsev** [comment]
> `name` in `moon:options` from previous comment, thanks!

@akrigator Ok, but it's inside `Configuration.browserCapabilities`.

**Comment by BorisOsipov** [comment]
>Yes, I am going to implement it. But can you specify what settings from Configuration you want to change from test to test (except Configuration.browserCapabilities)?

@asolntsev
`Configuration.browser`
`Configuration.assertionMode`

**Comment by asolntsev** [comment]
@BorisOsipov `browser` - yes. Probably also `remote`, `browserSize`, `browserPosition`, `holdBrowserOpen`, `

But why `assertionMode`? Isn't [1453](https://github.com/selenide/selenide/issues/1453) a better solution?

**Comment by vlad8x8** [comment]
are you going to create a monster to avoid a simple custom classes like 
```
class MyWebDriver extends SelenideDriver {
    public MyWebDriver() {
        super(new SelenideConfig()
                .headless(HEADLESS_FLAG)
                .browserCapabilities(WEB_OPTIONS));
    }
}
```

**Comment by asolntsev** [comment]
@vlad8x8 Definitely, I am not going to create a monster. I hope it will be a simple solution. 
But how `MyWebDriver` solves the problem? How exactly are you going to create and store instances of `MyWebDriver`?

**Comment by vlad8x8** [comment]
I store it as variables like 
`MyChromeDriver chrome = new MyChromeDriver();`
`MyFFDriver firefox = new MyFFDriver();`
then in the code, I can clearly understand what driver I'm going to use:
`chrome.$("button").click()`
`firefox.$("button").click()`

**Comment by asolntsev** [comment]
@vlad8x8 Yes, it may be convenient if you have 2+ browsers in a single test.
But in most cases, people run just 1 browser per test. They don't want to bother with `SelenideDriver` instances. They just want to use old good `$` in tests.
But they need to have different capabilities in different tests.

**Comment by AlexLAA** [comment]
@asolntsev 

Maybe just implement Configuration with ThreadLocal? People still will have possibility to use simple `$` in their tests. 
Browser & Configuration will be separated for each thread. Just need to implement it carefully with backward compatibility.

Our team also struggle with static Configuration case)
We need to run some tests in different browsers in parallel in one runtime.

**Comment by asolntsev** [comment]
@AlexLAA Maybe :)  It's not simple at all. Tons of nuances. 
And it will definitely cause many new errors & misunderstandings. I am pretty sure users will set Configuration in one thread, but use in another. And we will get lots of complains. 

Can you tell us **why** you need different browsers in parallel? 
I would say it's a wrong usage of parallelism if you ask me. :)

**Comment by AlexLAA** [comment]
I saw your opinion about it) Also I understand that it's not so simple to implement it in existing project)

**Our Usage**
We have one Test Suite per each Service. So when we deploy some Service we run Regression Suite for specific service.
For some of our Services we should run small tests in different browsers. 

It's pretty comfortable to have 1 Test-Suite Results for 1 Service Deployment. (Not to make chains of different jobs in Jenkins or any other CI).

**P.S.**
Of course it depends on usage, but for now it looks more like technical restriction, not ideology policy.
For my opinion it should be more flexible in parallelism. 
Tests and Suites Design should be on users side.

**Comment by AlexLAA** [comment]
@asolntsev If you want I can try to implement this feature, but I need some time to explore selenide detailed)

**Comment by BorisOsipov** [comment]
@AlexLAA what do you think about https://github.com/selenide/selenide/issues/1372#issuecomment-1264482125 ? does that suit you?

>And it will definitely cause many new errors & misunderstandings. I am pretty sure users will set Configuration in one thread, but use in another. And we will get lots of complains.

I agree with Andrey. We are struggling to find a way to implement it in a way that is easy to use and minimizes the risk of mistakes.

> If you want I can try to implement this feature, but I need some time to explore selenide detailed)

I guees you can pick up this issue and try. In any case, we will be grateful for your work and may potentially use it to implement a solution.

**Comment by asolntsev** [comment]
> Our Usage

I see your need would be covered by introducing method `open(url, options)`. 
I think it's much simpler and stable solution than introducing threadlocal Configuration analogue. 

> it looks more like technical restriction, not ideology policy.

No. It is technical restriction, but it was initially caused by clear understanding how should test automation (and generally software development) be organised.

**Comment by AlexLAA** [comment]
I think `open(url, options)` will cover it, just not sure about all possible cases with parallelism. It could be really great workaround.

The main point - it should be flexible. 
I understand your idea, but I cannot agree that selenide should declare the way test automation should be organised, it is each specific team responsibility.
It's not a process framework, it's useful tool for simplify writing tests via browser.

**Comment by asolntsev** [comment]
> It's not a process framework, it's useful tool for simplify writing tests via browser.

Generally I agree. Our goal is to simplify writing tests.
But the problem is that often tools do exactly the opposite: they make writing tests more complicated. Look at TestNG and Cucumber for example. Authors of such tools tend to add as many features as possible instead of simplifying. I don't want to create a monster. :)

I also think that it's a bad idea to introduce complicated features trying to find a technical solution for a non-technical problem. Such problems might be solved much more easily by fixing the process.

**Comment by ilya-corp** [comment]
In the my first project I need to create and manage all webdrivers manually by custom BrowserManager because my project requires simultaneous using of different browsers, for example there are driver for Firefox and custom chromedriver version 92 for Chrome to organize Electron app testing.

But the most important fault is inability to use proxy inthe  testA and not using proxy in testB **in the same suite**. 
Proxy is a global value of Configuration and I can't **safely** change it during the test. For example, I don't need proxy in 990 tests of 1000 and only 10 tests requires proxy.

**Comment by asolntsev** [comment]
@ilya-corp The idea is that:
1. if proxy works for you - enable it for all tests.
2. if proxy doesn't work for you (because of your network setup) - you cannot enable it anyway.

There is no reason to enable proxy only for few tests. 

P.S. Yes, I know, proxy breaks your tests when you try to download too large files (2+ GB). But it's rather a good reason to review your test design and avoid downloading so big files.

**Comment by sleepstream** [comment]
> @ivandros I realized that this feature may be implemented much simpler than I thought before.
> 
> All we need is to add parameter `browserCapabilities` to method `open(url)`. Then you will just need to call
> 
> ```java
> String videoIdentification  = currentClassName + "_" + new SimpleDateFormat("yyyy-MM-dd_HH:mm:ss").format(new Date()).mp4";
> 
> ChromeOptions chromeOptions = new ChromeOptions();
> chromeOptions.setCapability("name", videoIdentification);
> chromeOptions.setCapability("videoName", videoIdentification);
> open("", chromeOptions);
> ```
> 
> Right?
> 
> The only question: what should do method `open("", chromeOptions)` if there is already an opened browser from a previous test? Throw an exception? Or just log a warning? @BorisOsipov @bereg2k

Hi, is it already released?

**Comment by sleepstream** [comment]
> @sleepstream Yes, it's almost as simple as you showed.
> We can just close the previous browser when `open` is called with parameter `Capabilities`.
> 
> BUT
> We need to refactor some internals of Selenide because `WebDriverRunner.driver().config().browserCapabilities()` will still return the old capabilities (more exactly, the "global" value `Configuration.browserCapabilities`).

@asolntsev I don't understand....
Method open with Capabilities already available, from which version?

And what about refactor...

**Comment by asolntsev** [comment]
No, it's not available.
Current issue status is "OPEN".
